### PR TITLE
Introduced background reader, again

### PR DESCRIPTION
**Pull request by jstroem**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -137,32 +134,25 @@ export class EspLoader {
...
-        await this.read(false, 1000);
-      } catch (e) {
-        if (e === TimeoutError) {
-          break;
-        }
-      }
-      await sleep(50);
+    const wasSilent = await this.reader.waitSilent(20, 1000);
+    if (!wasSilent) {
+      this.options.logger.debug("try_connect reader was not silent");
```
Comment (jstroem) [inline src/index.ts:151]: I wonder if we should retry the bootloader startup step if this happens?

Comment (andersjohnsen) [reply src/index.ts:151]: Yes! Absolutely. Just a 5.repeat kind of thing

Comment (andersjohnsen) [reply src/index.ts:151]: IMO we should return here and say "Failed to enter bootloader". That could help us help the customer... Think GPy boards where they manually have to do something.

File: src/reader.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  public async read(minLength: number, timeoutMs: number): Promise<Uint8Array> {
+    if (this.listenRef <= 0) {
+      throw NotListeningError;
+    }
+    await this.waitData(minLength, timeoutMs);
+
+    return this.buffer.view(true);
+  }
+
+  public async packet(minLength: number, timeoutMs: number): Promise<Uint8Array> {
```
[the remark this answers is not in the document]
Comment (jstroem) [reply src/reader.ts:167]: yes, we have one scenario where we want to read length 6 (after memFinish)

File: src/index.ts (hunk 1 above)
Comment (jstroem) [reply src/index.ts:151]: I can write it out.

**Pull request merged**